### ISSUE TITLE
Data Source search: saved searches

### ISSUE THREAD

**Issue by josh-chamberlain** [Police-Data-Accessibility-Project/data-sources-app#21]
We should have a way to create these in the back end, so this can exist before [advanced search](#209) and without users.

Saved searches are the building blocks of a lot of things!

## Requirements
- [ ] using the quick search page should create a saved search and take people to the results page with the search ID, rather than `quick-search/${this.searchTerm}/${this.county}`
- [ ] a button on search results "save search" which makes a name automatically and takes the user to the new URL
- [ ] a "rename" button
- [ ] a "share" button which copies the link to the user's clipboard, or does more stuff on mobile
- [ ] allow people to see their saved searches
- [ ] saved searches must have permalinks, e.g. `https://pdap.io/search/UID123`
   - these save the search query itself, not the exact results of a search at a particular time
   - unless a good argument exists otherwise, a permalink like this is better than URL parameters
- [ ] this can be one table, with properties like `search_id`, `user_id`, `timestamp`, and `search_filters`
   - if we need to make it two tables later on, the migration should be pretty easy (i.e. one table for the queries, one table for the filtersets). For now, since searches and filtersets are 1:1, we should be good with one table and the benefits of multiple tables are probably minimal.

**Comment by maxachis** [comment]
> * [ ]  we need an endpoint to get saved and recent searches by user

@josh-chamberlain I want to make sure I understand what is meant by 'saved _and_ recent", since as previously mentioned, all user searches are considered saved. When we say recent, how recent are we talking? Are we talking within a time frame, or the last *x* searches?

**Comment by josh-chamberlain** [comment]
@maxachis 

1. the permalink is just the UID, so we should be able to generate it on the fly. Any reason we should store it?
2. a saved search ≠ its results; if you visit a saved search in a year, chances are the results will be different. 

all searches are saved ("logged" in a way), but when we talk about _a user's saved searches_, we're talking about searches they have _manually_ saved.

"recent" can mean the last, say, 10. we can change our definition of "recent" over time if we want.

**Comment by maxachis** [comment]
@josh-chamberlain 

> 1. the permalink is just the UID, so we should be able to generate it on the fly. Any reason we should store it?

I think that depends on how the saved queries work. I figured the search UIDs would be used to perform a lookup of the search in a database, retrieve the search parameters, and rerun them. Otherwise, how are we matching the UIDs with the particular search? And from where will the UIDs be generated? 

> the format of the saved query is up for debate, I think—and should be somewhat future-facing. see https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/14 for a more complete idea of all the things filters could contain. rather than a column for each search facet, maybe the query property should just be a JSON object which stores key: value pairs for different properties.

A JSON object might be the better option if we're intending to simply retrieve the search query and re-execute, rather than perform analyses on the parameters within the query itself (for example, how many people save queries for "Pittsburgh" as opposed to "Cincinnati"). The main challenges would lie in ensuring data integrity and backwards compatibility -- if our search logic changes at some point, we'd want to know if the previously saved searches are still valid (and if not, what to do with them -- perhaps convert them to a new format). 

At some point in the future, MongoDB or similar NoSQL databases might be useful to consider, as those are designed for storing information with flexible schemas, such as JSON. However, that might be overkill at the current moment, especially since our needs for what to do with the queries is minimal. However, may be worth keeping in the back of our minds.

**Comment by maxachis** [comment]
> 2. a saved search ≠ its results; if you visit a saved search in a year, chances are the results will be different.

@josh-chamberlain In which case, I may need clarification on what is meant by 

> we should keep track how many saved searches a given source is part of; sounds like an expensive query?

If I'm understanding this correctly, and we are looking to track how often a given search is returned in queries, we could create a table which increments every time a data source is retrieved in a search/saved search.

**Comment by josh-chamberlain** [comment]
> I think that depends on how the saved queries work. I figured the search UIDs would be used to perform a lookup of the search in a database, retrieve the search parameters, and rerun them. Otherwise, how are we matching the UIDs with the particular search? And from where will the UIDs be generated?

yeah, I think we're saying the same thing—I meant we don't need to store the url if we just know to use that route, but I'm splitting hairs 🤷

> A JSON object might be the better option if we're intending to simply retrieve the search query and re-execute, rather than perform analyses on the parameters within the query itself (for example, how many people save queries for "Pittsburgh" as opposed to "Cincinnati"). The main challenges would lie in ensuring data integrity and backwards compatibility -- if our search logic changes at some point, we'd want to know if the previously saved searches are still valid (and if not, what to do with them -- perhaps convert them to a new format).

yes, I think logging is a separate concern (I think)

At some point in the future, MongoDB or similar NoSQL databases might be useful to consider, as those are designed for storing information with flexible schemas, such as JSON. However, that might be overkill at the current moment, especially since our needs for what to do with the queries is minimal. However, may be worth keeping in the back of our minds.

**Comment by josh-chamberlain** [comment]
> I think that depends on how the saved queries work. I figured the search UIDs would be used to perform a lookup of the search in a database, retrieve the search parameters, and rerun them. Otherwise, how are we matching the UIDs with the particular search? And from where will the UIDs be generated?

yeah, I think we're saying the same thing—I meant we don't need to store the url if we just know to use that route, but I'm splitting hairs 🤷

> A JSON object might be the better option if we're intending to simply retrieve the search query and re-execute, rather than perform analyses on the parameters within the query itself (for example, how many people save queries for "Pittsburgh" as opposed to "Cincinnati"). The main challenges would lie in ensuring data integrity and backwards compatibility -- if our search logic changes at some point, we'd want to know if the previously saved searches are still valid (and if not, what to do with them -- perhaps convert them to a new format).

yes, I think logging is a separate concern. Backwards compatibility is definitely a consideration, but if we're keeping logs, I think it's OK to change the format. [I made a new issue](https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/270).

> At some point in the future, MongoDB or similar NoSQL databases might be useful to consider, as those are designed for storing information with flexible schemas, such as JSON. However, that might be overkill at the current moment, especially since our needs for what to do with the queries is minimal. However, may be worth keeping in the back of our minds.

Absolutely—we could also use it to dump `actual data` if we stick some semi-predictable metadata to it and hook it up to elasticsearch.

> If I'm understanding this correctly, and we are looking to track how often a given search is returned in queries, we could create a table which increments every time a data source is retrieved in a search/saved search.

That's a good idea! If we can't do it in a simple way it's probably not worth worrying about. I would call it a nice-to-have. Simpler, maybe just add a column? and `in_searches` `last_in_search` datetime. We could also just make that part of logging and keep it out of the db.

**Comment by maxachis** [comment]
@josh-chamberlain So next question is related to the form of the UUID. 

One thing I'll also need clarification on -- does a saved search, when accessed, merely provide a suite of results? Will anyone other than the user themselves know which saved searches are associated with them? Because that will affect how concerned we are about issues like a hacker trying to enumerate through all saved searches.

The way I see it, there are two options, and one option we definitely don't want to do.

### 🚫Serial IDs 🚫
Serial IDs can be guessed or enumerated through, meaning that an individual could use them to effectively find every single search in our database, or to identify a user associated with a suite of saved searches by cross-referencing with a number of saved searches clearly related to each other by the proximity of their serial ids. Bad.

### Simple UUID
This is the easiest solution: We generate a random 128-bit UUID which is almost certainly mathematically impossible to ever be made again ever. This would look like `550e8400-e29b-41d4-a716-446655440000`. Simple, but not exactly easy to share.

### Bitly-style short ID
Generate a random short ID like `2KEOXNx`. Much easier to share, however, has some issues:

1. Risk of generating the same ID twice increases -- still generally quite small, but we have to figure out how to handle them if and when they occur, which would become an increasing issue as the number of saved searches increases.
2. Increased risk of enumeration attacks: i.e., an enterprising hacker could try to generate all possible combinations to get access to all searches (or, alternatively, to overload the system by generating a bunch of saved searches at once). Can be averted with rate limiting and/or limiting the number of saved searches a user can have, but either way, more overhead.

**Comment by maxachis** [comment]
@josh-chamberlain In which case, serial ids won't be a problem! I'll utilize serial ids as the solution.

**Comment by josh-chamberlain** [comment]
@maxachis as I thought may happen, there are some changes to the plan which will make `saved searches` unimportant until we build `advanced search filtering`—which is maybe the least launch-critical feature—because we are going to stick to a URL format for search. People will only need to _save searches_ if they make a search which can't be expressed via the URL, which they won't be able to do.

**Comment by maxachis** [comment]
@josh-chamberlain In that case, I'll close the pull request I made for this just to avoid clutter. We can still use its material (if it's still relevant) when the time comes.

**Comment by maxachis** [comment]
@josh-chamberlain Since this is nearly a year old, is this still an issue we want to do?

**Comment by josh-chamberlain** [comment]
@maxachis no, the "followed searches" concept more or less replaces this.
